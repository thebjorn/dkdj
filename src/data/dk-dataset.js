
import dk from "../dk-obj";
import {DataPage} from "./dk-datapage";
import Class from "../lifecycle/coldboot/dk-class";
import {DataQuery} from "./dk-dataquery";
import {ArraySource} from "./source/dk-array-datasource";
import {is_ajax_url} from "../lifecycle/uri";
import {AjaxDataSource} from "./source/dk-ajax-datasource";


/*
 *  A DataSet is an abstract paged collection of data items.
 *
 *  A DataSet can provide a conrete DataPage (i.e. a list of
 *  records to be displayed).
 *
 *  The purpose of a DataSet is to provide such
 *  a DataPage by querying the underlying data source.
 *
 *  DataSet -> DataPage -> (data) Record -> (data) Field) -> (data) FieldValue
 */
export class DataSet extends Class {
    constructor(props) {
        if (props == null) props = {};
        if (Array.isArray(props.datasource)) props.datasource = new ArraySource(props.datasource);
        if (is_ajax_url(props.datasource)) props.datasource = new AjaxDataSource({url: props.datasource});
        super(Object.assign({
            // type: 'dk.data.Data',
            datasource: null,
            pages: null,            // dict[query -> page]
            page: null,             // current page

            pagenum: 0,             // current page number
            totpages: 0,            // total number of pages

            pagesize: 25,
            orphans: 0,
            
        }, props));
    
        this._first_fetch = true;
        this.pages = {};
        this._filter_data = {};
    }

    // synchronize dirty elements on page to datasource.
    update() {
        this.pages = {};  // clear cache
        this.page.update(this.datasource, this.FN('_new_recordset'));
        // this.page.update(this.datasource, function (recordset) {
        //     self._new_recordset(self._get_records(null, recordset));
        // });
    }

    /**
     * Gets the current state
     * @param query  - a DataQuery instance
     * @returns {*}  - a DataQuery instance
     */
    get_state(query) {
        if (!this.datasource.url) return query;
        return DataQuery.create(dk.hash.get(this.datasource.url, query.copy()), this);
    }

    set_state(query) {
        if (this.datasource.url) {
            dk.hash[this.datasource.url] = query.copy();
        }
    }
    
    /**
     * _new_recordset is the callback handler for this.datasource.get_records().
     * 
     * @param recordset
     * @param query
     * @private
     */
    _new_recordset(recordset, query) {
        const page = DataPage.create({
            query: query,
            recordset: recordset
        });
        if (recordset.meta) {
            const m = recordset.meta;
            //dk.dir('query', query);
            if (m.totcount === undefined && m.filter_count === undefined) {
                m.totcount = 0;
                m.filter_count = 0;
            }
            if (m.totcount && m.filter_count === undefined) m.filter_count = m.totcount;
            if (m.pagecount === undefined && m.filter_count === 0) {
                m.pagecount = 0;
            } else if (m.filter_count !== undefined && !m.pagecount) {
                m.pagecount = Math.max(1, Math.floor(m.filter_count / query.pagesize));
                if (m.filter_count > query.pagesize && m.filter_count % query.pagesize > query.orphans) {
                    ++m.pagecount;
                }
            }
        }
        dk.on(page, 'dirty', (...args) => this.update(...args));
        this.page = this.pages[query] = page;
        dk.trigger(this, 'fetch-info', page.recordset.meta, query);
        dk.trigger(this, 'fetch-data', this, page);
    }

    get_page(query) {
        query = DataQuery.create(query, this);
        // console.info("DATASET:GET:PAGE:QUERY:", this._first_fetch, query);
        if (this._first_fetch) {
            query = this.get_state(query);
            this._first_fetch = false;
        } else {
            this.set_state(query);
        }
        // console.info("DATASET:GET:PAGE:QUERY:2", this._first_fetch, query);
        
        // dk.dir("GET-PAGE:", query.copy());
        dk.trigger(this, 'fetch-data-start');

        if (!this.pages[query]) {
            // this.datasource.get_records(query, this.FN('_new_recordset'));
            this.datasource.get_records(query, recordset => {
                this._new_recordset(recordset, query);
            });
        } else {
            dk.info("GET-PAGE... cached!");
            this.page = this.pages[query];
            dk.trigger(this, 'fetch-info', this.page.recordset.meta, query);
            dk.trigger(this, 'fetch-data', this, this.page);
        }
    }

    _current_query() {
        if (this.page) return this.page.query.copy();
        return {};
    }

    /*
     *  Fetch page number `n` (zero-based).
     */
    get_pagenum(n) {
        const query = this._current_query();
        query.pagenum = n;
        this.get_page(query);
    }

    set_search(terms) {
        const query = this._current_query();
        query.search = terms;
        query.pagenum = 0;
        this.get_page(query);
    }

    set_sort(fieldname, direction) {
        const query = this._current_query();
        const newsort = {field: fieldname, direction: direction};
        query.sort = [newsort];
        query.pagenum = 0;
        this.get_page(query);
    }

    get_filter_data(filter_name, returns) {
        // get data needed to display filter with name `filter_name`
        if (this._filter_data[filter_name]) {
            returns(this._filter_data[filter_name]);
        } else {
            this.datasource.get_filter_data(filter_name, opts => {
                returns(opts);
            });
        }
    }

    set_filter(vals) {
        const query = this._current_query();
        query.filter = vals;
        query.pagenum = 0;
        this.get_page(query);
    }

    fetch_csv(filename) {
        let query = this._current_query();
        query.pagesize = 50000;
        query = DataQuery.create(query, this);
        if (dk.$('#downloadFile').length) {
            dk.$('#downloadFile').remove();
        }
        const url = this.datasource.url+"!get-records?fmt=csv&filename="+ filename + '&' + query.toGetParams();
        dk.$('<a></a>')
            .attr('id','downloadFile')
            .attr('href',url)
            .attr('target', '_blank')
            .appendTo('body');
        dk.$('#downloadFile').ready(function() {
            dk.$('#downloadFile').get(0).click();
        });
    }
}
