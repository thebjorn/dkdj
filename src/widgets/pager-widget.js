
// const $ = require('jquery');
// const dk = require('../boot');
// const Widget = require('../widgetcore/dk-widget.js');

import dk from "../dk-obj";
import {Widget} from "../widgetcore/dk-widget";
import {dkconsole} from "../lifecycle/dkboot/dk-console";

export class PagerWidget extends Widget {
    constructor(...args) {
        super({
            // type: 'PagerWidget',

            pagecount: 0,               // total number of pages
            curpage: undefined,         // currently selected page (zero-based)

            classes: ['pagination'],
            css: {
                marginTop: 0,
                marginBottom: 0
            },
            template: {root: 'ul'},
            
        }, ...args);
    }
    

    /*
     *  Set the pagecount to `n`.
     */
    set_pagecount(n) {
        if (this.pagecount !== n) {
            this.pagecount = n;
            this.draw();
        }
    }

    /*
     *   Select page `n` (zero-based).
     */
    select_page(n) {
        if (n === this.curpage || this.pagecount=== 0) return;     // nothing to do
        if (n < 0 || n >= this.pagecount) {
            dkconsole.debug("PagerWidget.select_page: out of bounds ", n, this);
            return;
        }
        this.curpage = n;
        this.draw();
        this.trigger('select-page', n);
    }

    handlers() {
        const self = this;
        this.widget().on('click', 'li.page', function () {
            self.select_page(dk.$(this).data('page'));
        });
        this.widget().on('click', 'li.prev-nav', function () {
            if (self.curpage > 0) {
                self.select_page(self.curpage - 1);
            }
        });
        this.widget().on('click', 'li.next-nav', function () {
            if (self.curpage + 1 < self.pagecount) {
                self.select_page(self.curpage + 1);
            }
        });
    }
    _prerange(a, b) {
        return (b < a)? null: [a, Math.min(b + 1, a + 2)];
    }
    _postrange(a, b) {
        return (a > b)? null: [Math.max(a, b - 1), b];
    }
    draw_range(range) {
        for (let i = range[0]; i <= range[1]; i++) {
            const li = dk.$('<li/>', { page: i }).addClass('page').data('page', i-1);
            li.append(dk.$('<a href="#"/>').text(i));
            if (i === this.curpage + 1) {
                li.addClass('active');
            }
            this.widget().append(li);
        }
    }
    draw() {
        this.widget().empty();
        this.widget().append(dk.$('<li class="prev-nav"><a href="#">&laquo;</a></li>'));

        if (this.pagecount > 0) {
            let prerange = this._prerange(1, this.curpage - 3);
            let midrange = [Math.max(1, this.curpage - 3), Math.min(this.pagecount, this.curpage + 3)];
            let postrange = this._postrange(this.curpage + 3, this.pagecount);

            // console.log("PRE:", prerange);
            // console.log("MID:", midrange);
            // console.log("POST:", postrange);
            // console.log('----after...---');
            
            if (prerange && prerange[1] + 1 >= midrange[0]) {
                midrange[0] = 1;
                prerange = null;
            }
            if (postrange && midrange[1] + 1 >= postrange[0]) {
                midrange[1] = postrange[1];
                postrange = null;
            }
            // console.log("PRE:", prerange);
            // console.log("MID:", midrange);
            // console.log("POST:", postrange);
            if (prerange) {
                this.draw_range(prerange);
                this.widget().append(dk.$('<li class="ellipsis"><span>...</span></li>'));
            }
            if (midrange) {
                this.draw_range(midrange);
            }
            if (postrange) {
                this.widget().append(dk.$('<li class="ellipsis"><span>...</span></li>'));
                this.draw_range(postrange);
            }
        }

        this.widget().append(dk.$('<li class="next-nav"><a href="#">&raquo;</a></li>'));
    }

}
