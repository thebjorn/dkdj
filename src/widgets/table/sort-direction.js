
// var Widget = require('../../widgetcore/dk-widget.js');


import {Widget} from "../../widgetcore/dk-widget";
import {icon} from "../dk-icon-library";

export class SortDirection extends Widget {
    constructor(...args) {
        super({
            direction: '',  // asc/desc/''
            width: 20,
            structure: {
                classes: ['sort-icon'],
                css: { float: 'right' }
            }
        }, ...args);
        this.direction = '';
        this.width = this.width || 20;

    }
    
    construct() {
        this.widget().css({
            width: this.width,
            marginLeft: this.width
        });
    }

    set_direction(direction) {
        if (direction !== this.direction) {
            this.direction = direction;
            this.notify('sort', direction);
        }
    }

    clear_sort() {
        this.set_direction('');
        this.draw();
    }

    toggle_sort_direction() {
        this.set_direction({
            asc: 'desc',
            desc: 'asc',
            '': 'desc'
        }[this.direction]);
        this.draw();
    }

    handlers() {
        var self = this;
        self.widget().on('click', function () {
            self.toggle_sort_direction();
        });
    }

    draw() {
        this.widget().empty();
        switch (this.direction) {
            case 'asc':
                this.widget().append(icon('up'));
                break;
            case 'desc':
                this.widget().append(icon('down'));
                break;
        }
    }
}
