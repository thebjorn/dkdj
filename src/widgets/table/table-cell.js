

import dk from "../../dk-obj";
import {Widget} from "../../widgetcore/dk-widget";


/*
 *  The TableCell class is responsible for pushing and pulling data into a
 *  table cell.  The data itself is contained in the enclosing TableRow's
 *  `record` property.
 */
export class TableCell extends Widget{
    constructor(...args) {
        super({
            tablerow: null,
            coldef: null,
        }, ...args);
    }

    set value(val) {
        if (!this.tablerow) return;
        this.tablerow.set_column_value(this.coldef, val);
    }
    get value () {
        if (!this.tablerow) return null;
        return this.tablerow.get_column_value(this.coldef);        
    }

    construct_edit_widget() {
        const location = this.widget();
        const coldef = this.coldef;
        const value = this.value;

        let width = location.width();
        location.empty();
        const edit_widget = coldef.widget_type.create_inside(location, {
            data: coldef.widget_data || {},  // eg. data to fill a drop-down list.
            url: coldef.widget_url,    // url to fetch data from
            css: { border: 'none' }
        });

        if (coldef.widget_type === 'SelectWidget') width += 20;  // XXX: yeuch!
        edit_widget.widget().width(width);
        edit_widget.value = value;  // does the right thing for select too..
        return edit_widget;
    }

    start_editing() {
        // prevent infinite recursion
        if (this.editing || this.coldef.editable === false) return;
        this.editing = true;

        if (!this._undo_stack) this._undo_stack = [];
        this._undo_stack.push(this.value);              // save current value, in case of undo.

        this.edit_widget = this.construct_edit_widget();

        // update field immediately on change (we can still undo..)
        this.edit_widget.widget().on('change', () => {
            const oldval = this.value;
            this.value = this.edit_widget.get_field_value();
            dk.trigger(this, 'dirty', this.coldef, this.value, this.edit_widget, oldval);
        });
    }

    save_editing() {
        if (this.coldef.editable === false) return;
        this.value = this.edit_widget.get_field_value();
        this._undo_stack.pop();
    }

    undo_editing() {
        if (this.coldef.editable === false) return;
        this.value = this._undo_stack.pop();
    }

    stop_editing() {
        if (this.coldef.editable === false) return;
        this._undo_stack = [];
        this.widget().empty();
        this.draw();
        this.editing = false;
    }

    draw(value) {
        const val = this.value === undefined ? value: this.value;
        this.widget().attr({
            axis: this.coldef.name,
            title: this.coldef.title
        }).html(this.coldef.format(val, this.tablerow.record, this.widget()));
    }
}
