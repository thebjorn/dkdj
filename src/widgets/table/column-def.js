
import Class from "../../lifecycle/coldboot/dk-class";
import dk from "../../dk-obj";
import {value as format_value} from "../../data/datacore/dk-format"
import widgetmap from "../../widgetcore/dk-widgetmap";
import {TextInputWidget} from "../../forms/widgets";


export class ColumnDef extends Class {
    constructor(...args) {
        super({
            isa:         'dk.table.ColumnDef',
            name:        '',                        // name of column (eg. for computed cols)
            field:       '',                        // field definition {name:.., pos:.., string:..,...}
            label:       '',                        // text to put in th element
            sortable:    undefined,                 // sortable override
            description: '',                        // description
            align:       undefined,                 // cell alignment
            format:      format_value,  // formatting function
            type:        undefined,                 // data type of column values
            widget_type: TextInputWidget,     // widget to use when editing a cell value
            empty:       '',                        // value to use for empty cells
            table:       undefined,                 // the DataTable (subclass) we belong to
            _alignment_map: {
                int: 'right'
            }
        }, ...args);
        this.name = this.name || this.field;
    }

    alignment() {
        return this.align || this._alignment_map[this.type] || 'left';
    }

    get_value(record) {
        //dk.info("getting value from columndef", record, this, record[this.name]);
        if(record)
            return record[this.name];
    }

    /*
     *  Update self with info from the datasource.
     */
    bind_to_field(field) {
        this.field = field;
        if (!this.label) this.label = field.label || field.name;
        if (!this.type) this.type = field.type;

        let wtype;
        if (field.widget) {
            // wtype = widgetmap.get(field.widget);
            wtype = widgetmap[field.widget];
        } else {
            wtype = widgetmap[this.type];
        }

        if (wtype === undefined) {
            // dk.warn("Unknown widget: ", field.widget, field);
            wtype = TextInputWidget;
        }
        this.widget_type = wtype;
        this.widget_data = field.data;
        this.widget_url = field.url;
    }
}
