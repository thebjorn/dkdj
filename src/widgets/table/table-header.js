

import dk from "../../dk-obj";
import {TableCell} from "./table-cell";
import {cursor} from "../cursors";
import {SortDirection} from "./sort-direction";

/*
 *  Column header widget.
 */
export class TableHeader  extends TableCell {
    constructor(...args) {
        super({
            scope: 'col',
            width: undefined,
            sortable: true,

            name: undefined,
            label: undefined,
            title: undefined,
            format: undefined,
            field: undefined,
            table: undefined,

            template: {root: 'th'},

        }, ...args);
    }
    

    construct() {
        if (this.sortable) {
            this.sort_icon = SortDirection.create_inside(this.widget(), {
                width: '2ex'
            });
        }
        this.title = this.description;
    }

    handlers() {
        const self = this;
        if (this.sortable) {
            self.widget().on('click', function () {
                self.table.set_sort(self);
            });
            dk.on(this.sort_icon, 'sort', () => this.set_sort_cursor());
        }
    }

    set_sort_cursor() {
        if (this.sortable) {
            const _cursor = this.sort_icon.direction === 'desc' ? 'down' : 'up';
            this.widget().css('cursor', cursor(_cursor));
            // this.table.notify('sort', this.sort_icon.direction);
            // this.table.trigger('sort', this.sort_icon.direction);
            dk.trigger(this.table, 'sort', this.sort_icon.direction);
        }
    }

    draw() {
        this.widget().attr({
            title: this.title,
            scope: this.scope,
            fieldname: this.name
        }).append(this.label);
        if (this.sortable) this.set_sort_cursor();
    }
}
