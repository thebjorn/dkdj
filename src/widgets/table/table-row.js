
/*
 *  A Data Record widget.
 */

import {Widget} from "../../widgetcore/dk-widget";
import {TableCell} from "./table-cell";
import {TableRowLayout} from "../../layout/table-layout";


export class TableRow extends Widget {
    constructor(...args) {
        super({
            dklayout: TableRowLayout,
            rownum: null,
            record: null,
            table: null,
            cells: [],
            template: {root: 'tr'},
        }, ...args);
    }
    
    delete_widget() {
        delete this.record;
        delete this.table;
        this.cells.forEach(c => c.delete_widget());
        this.cells = [];
        super.delete_widget();
    }

    construct() {
        const record = this.record;
        const table = this.table;
        this.widget().attr({
            pk: record.pk,
            rownum: this.rownum
        });
        this.cells = table.column_order.map(fname => {
            const coldef = table.column[fname];
            return TableCell.create_on(this.layout.add_td(), {
                tablerow: this,
                coldef: coldef
            });
        });
    }

    get_column_value(coldef) {
        return coldef.get_value(this.record);
    }
    set_column_value(coldef, val) {
        this.record[coldef.name] = val;
    }
}
