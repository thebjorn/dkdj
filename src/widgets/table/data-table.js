/*
 *  A DataTable is the normal table representation of a DataSource,
 *  i.e. records in rows, with each field in its own column (with a column
 *  header).
 */

import dk from "../../dk-obj";
import {TableLayout} from "../../layout/table-layout";
import {Widget} from "../../widgetcore/dk-widget";
import {TableRow} from "./table-row";
import is from "../../is";
import {ColumnDef} from "./column-def";
import {TableHeader} from "./table-header";
import {DataSet} from "../../data/dk-dataset";
import {AjaxDataSource} from "../../data/source/dk-ajax-datasource";
import {dkwarning} from "../../lifecycle/coldboot/dkwarning";


export class DataTableLayout extends TableLayout {
    constructor(widget, location, template, structure) {
        super(widget, location, template, structure);
        this.thead_tr = this.add_header_row();
    }
    
    add_thead_th(attrs) {
        const th = this.make('thead_th', attrs);
        this.thead_tr.appendln(th);
        return th;
    }
}


export class DataTable extends Widget {
    constructor(...args) {
        super({
            // type: 'datatable',
            dklayout: DataTableLayout,
            TableRow: TableRow,
            current_pagenum: 0,
            translate: {},  // l10n

            structure: {
                thead: {},
                tfoot: {},
                tbody: {}
            },

            template: {
                root: 'table',
                table: '<table/>',
                thead: '<thead/>',
                thead_tr: '<tr/>',
                thead_th: '<th/>',
                tbody: '<tbody/>',
                tbody_tr: '<tr/>'
            },

            table_data: null,
            state: undefined,

            // The columns attribute should be a property object that defines the visible columns:
            //
            //      columns: {
            //          colnam1: {}
            //          colname2: {..}
            //      ..
            //
            // this.columns is converted to an array of dk.table.ColumnDef, and this.column to a
            // dict from column name to dk.table.ColumnDef.
            columns: null,
            column: {},

            filters: {},

            // should table be sortable?
            sortable: true,

            // selector of dom element to bless as a download link (e.g.
            // <a id="foo" data-download="filename.csv">download</a>
            // =>  download: '#foo',
            download: undefined,

            header: null,
            rows: null,

            column_order: null,
            column_defs: null,            
        }, ...args);

        const self = this;
        this.rows = [];
        this.header = [];   // th cell objects

        this.column = {};
        this.column_order = [];
        if (!!this.datasource && this.table_data === null) {
            const dsprops = {datasource: this.datasource};
            if (this.pagesize != null) dsprops.pagesize = this.pagesize;
            if (this.orphans != null) dsprops.orphans = this.orphans;
            if (this.pagenum != null) dsprops.pagenum = this.pagenum;
            this.table_data = new DataSet(dsprops);
        }
        if (this.table_data === null) {
            this.data_url = (this.dk && this.dk.url)? this.dk.url : "";
            //dk.info("URL", this.data_url);
            // if data is not defined, then create one
            this.table_data = new DataSet({
                // default data source is AjaxDataSource
                datasource: new AjaxDataSource({
                    // default url is current url
                    url: self.data_url
                })
            });
        }
        if (!this.data_url) this.data_url = this.table_data.datasource.url;

        //        this.state = dk.State.create({
        //            state: this.table_data.current_pagedef,
        //            engine: dk.HashStorage.create({})
        //        });
        //this.state.restore();
    }

    construct() {
        // this.state = dk.page.hash.substate(this.id);

        if (this.download) {
            const $download = dk.$(this.download);
            const filename = $download.attr('data-download') || 'filename.csv';
            dk.$(this.download).attr('href', this.data_url + "!get-records?fmt=csv&filename=" + filename);
        }
    }

    /*
     *  Connect to a pager widget.
     */
    set_pager(pwidget) {
        // const self = this;
        if (this.table_data.page) {
            pwidget.set_pagecount(this.table_data.page.recordset.meta.pagecount);
            pwidget.select_page(this.table_data.page.query.pagenum);
        }

        dk.on(pwidget, 'select-page', (...args) => this.table_data.get_pagenum(...args));
        dk.on(this.table_data, 'fetch-info', (info, query) => {
            pwidget.set_pagecount(info.pagecount);
            pwidget.select_page(query.pagenum);
        });
    }

    filter_values() {
        return {};
    }

    handlers() {
        //$(this.table_data).on('fetch-data', _.bind(this.draw, this));
        //$bind('fetch-data@data -> draw@me', {data: this.table_data, me: this});
        dk.on(this.table_data, 'fetch-data-start', () => this.start_busy());
        dk.on(this.table_data, 'fetch-data', (...args) => this.draw(...args));
        dk.on(this.table_data, 'fetch-data', (...args) => this.end_busy(...args));
    }

    set_sort(sortcol) {
        if (!sortcol.sortable) return;
        if (this.sortable) {
            this.header.forEach(function (col) {
                if (col.sortable && col !== sortcol) col.sort_icon.clear_sort();
            });
            sortcol.sort_icon.toggle_sort_direction();
            this.table_data.set_sort(sortcol.name, sortcol.sort_icon.direction);
            dk.trigger(this, 'sort-order', sortcol.name, sortcol.sort_icon.direction, this);
        }
    }

    set_search(terms) {
        this.table_data.set_search(terms);
        if (this.state) this.state.set('pagedef', this.table_data.current_pagedef);
    }

    set_filter(vals) {
        this.table_data.set_filter(vals);
    }

    get_xy(colnum, rownum) {
        return this.rows[rownum].cells[colnum];
    }

    set_xy(colnum, rownum, val, html, title) {
        const cell = this.get_xy(colnum, rownum);
        cell.set_value(val, html, title);
        return cell;
    }

    values() {
        const res = [];
        this.rows.forEach(function (row) {
            const r = [];
            row.cells.forEach(function (cell) {
                r.push(cell.widget().text());
            });
            res.push(r);
        });
        return res;
    }

    /*
     *  Convert this.columns to an array of basic property objects.
     *
     *    {foo: {label: 'Foo', ..}} ==> [{name: 'foo', label: 'Foo', ...}, ...]
     */
    _get_columndefs(fieldlst) {
        const self = this;
        let columndefs;
        const fields = {};  // field.name -> field
        fieldlst.forEach(function (f) { fields[f.name] = f; });

        // make sure this.columns is an array of pojos
        if (is.isFunction(this.columns)) columndefs = this.columns(this.table_data.fields);
        else if (this.columns === null) columndefs = fieldlst;
        else if (!Array.isArray(this.columns)) {
            // the columns property has been defined, convert it to array
            columndefs = Object.keys(this.columns).map(name => {
                const col = this.columns[name];
                col.name = name;
                return col;
            });
        } else {
            dkwarning(`Expected {field1: {...}, field2: {...}, etc.}. columns property of unknown type: ${this.columns}`);
        }

        // convert the pojos to ColumnDef objects
        return columndefs.map(function (coldef, colnum) {
            const field = fields[coldef.name];
            const tmp = {};
            if (field) {
                Object.assign(tmp, {
                    fieldpos: field.pos,
                    fieldtype: field.type,
                    sortable: field.sortable,
                    label: field.label,
                    help_text: field.help_text,
                    field_widget: field.widget,
                    field_data: field.data
                });
            }
            Object.assign(tmp, coldef);
            Object.assign(tmp, {
                colnum: colnum,
                table: self
            });
            if (tmp.sortable === undefined) {
                tmp.sortable = self.sortable;
            }
            const cdef = new ColumnDef(tmp);
            if (fields[cdef.name]) cdef.bind_to_field(fields[cdef.name]);
            self.column[cdef.name] = cdef;
            self.column_order.push(cdef.name);
            return cdef;
        });
    }

    /*
     *  Draw the column headers.
     */
    draw_header(dataset) {  // must be a separate method so VDataTable can inherit..
        if (!this.drawn_header) {
            const self = this;
            // NOTE: we have data here!
            this.columns = this._get_columndefs(dataset.page.fields);

            this.columns.forEach(function (coldef) {
                const th = TableHeader.create(self.layout.add_thead_th(), coldef);
                self.header.push(th);
            });
            self.drawn_header = true;  // only draw header once.
        }
    }

    draw(dataset) {
        const self = this;

        if (!dataset) {   // NOTE: data source consumer pattern..?  (on: data -> draw?)
            this.table_data.get_page({
                pagenum: this.current_pagenum
            });
        } else {
            self.draw_header(dataset);
            dk.trigger(self, 'draw-start', self);
            self.layout.clear_body();
            self.rows = [];

            dataset.page.records.forEach(function (record, rownum) {
                const tr = self.TableRow.create(self.layout.add_row_to('tbody'), {
                    rownum: rownum,
                    record: record,
                    table: self
                });
                self.rows.push(tr);
            });
            dk.trigger(self, 'draw-end', self);
        }
    }
}
