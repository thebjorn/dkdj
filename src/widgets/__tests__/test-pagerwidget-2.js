
import $ from 'jquery';
import {PagerWidget} from "../pager-widget";
import page from "../../widgetcore/dk-page";
import {DataSet} from "../../data/dk-dataset";
import {DataTable} from "../table/data-table";
import {ArraySource} from "../../data/source/dk-array-datasource";


test("test-pagerwidget-with-datatable", () => {
    document.body.innerHTML = `
        <div id="work">
            <div class="container">
                <div class="dtbl"></div>
                <div class="pgr"></div>
            </div>
        </div>
    `;
    const work = $('#work');
    page.initialize(document);

    const dt = DataTable.create_inside(work.find('.container > .dtbl'), {
        table_data: DataSet.create({
            datasource: ArraySource.create([
                {project: 'first', work: '1:03:57'},
                {project: 'aTiktok', work: '2:44:57'},
                {project: 'bGenerelt NT', work: '1:03:57'},
                {project: 'cTiktok', work: '2:44:57'},
                {project: 'dGenerelt NT', work: '1:03:57'},
                {project: 'eTiktok', work: '2:44:57'},
                {project: 'fGenerelt NT', work: '1:03:57'},
                {project: 'gTiktok', work: '2:44:57'},
                {project: 'hGenerelt NT', work: '1:03:57'},
                {project: 'iTiktok', work: '2:44:57'},
                {project: 'jGenerelt NT', work: '1:03:57'},
                {project: 'kTiktok', work: '2:44:57'},
                {project: 'last', work: '1:06:43'}
            ]),
            pagesize: 5,
            orphans: 4
        }),

        columns: {
            project: {label: 'Prosjekt'},
            work: {label: 'Arbeid'}
        }
    });
    const pg = dt.table_data.page;
    expect(dt.table_data.page.recordset.meta.pagecount).toBe(2);
    const pager = PagerWidget.create_inside(work.find('.container > .pgr'), {});
    dt.set_pager(pager);
    pager.widget('[page=2]').click();
    console.info("PAGER:", pager);
    console.log(work.html());

    expect(document.getElementById('work')).toMatchSnapshot();
});
