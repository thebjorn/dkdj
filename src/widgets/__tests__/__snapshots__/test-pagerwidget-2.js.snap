// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test-pagerwidget-with-datatable 1`] = `
<div
  id="work"
>
  
            
  <div
    class="container"
  >
    
                
    <div
      class="dtbl"
    >
      <table
        class="DataTable dk-bx"
        id="data-table-1"
      >
        <thead
          id="dk-bx-1"
        >
          

          <tr
            id="dk-table-1"
          >
            

            <th
              class="string"
              fieldname="project"
              id="dk-table-2"
              scope="col"
              style="cursor: url(\\"https://static.datakortet.no/dk/sort-za.cur\\"), pointer;"
              title=""
            >
              <div
                class="SortDirection sort-icon"
                id="sort-direction-1"
                style="float: right; width: 2ex; margin-left: 2ex;"
              />
              Prosjekt
            </th>
            

            

            <th
              class="string"
              fieldname="work"
              id="dk-table-3"
              scope="col"
              style="cursor: url(\\"https://static.datakortet.no/dk/sort-za.cur\\"), pointer;"
              title=""
            >
              <div
                class="SortDirection sort-icon"
                id="sort-direction-2"
                style="float: right; width: 2ex; margin-left: 2ex;"
              />
              Arbeid
            </th>
            

          </tr>
          

        </thead>
        <tfoot
          id="dk-bx-2"
        />
        <tbody
          id="dk-bx-3"
        >
          

          <tr
            class="TableRow"
            id="dk-table-9"
            pk="5"
            rownum="0"
          >
            

            <td
              axis="project"
              class="TableCell"
              id="dk-tr-11"
            >
              eTiktok
            </td>
            

            

            <td
              axis="work"
              class="TableCell"
              id="dk-tr-12"
            >
              2:44:57
            </td>
            

          </tr>
          

          

          <tr
            class="TableRow"
            id="dk-table-10"
            pk="6"
            rownum="1"
          >
            

            <td
              axis="project"
              class="TableCell"
              id="dk-tr-13"
            >
              fGenerelt NT
            </td>
            

            

            <td
              axis="work"
              class="TableCell"
              id="dk-tr-14"
            >
              1:03:57
            </td>
            

          </tr>
          

          

          <tr
            class="TableRow"
            id="dk-table-11"
            pk="7"
            rownum="2"
          >
            

            <td
              axis="project"
              class="TableCell"
              id="dk-tr-15"
            >
              gTiktok
            </td>
            

            

            <td
              axis="work"
              class="TableCell"
              id="dk-tr-16"
            >
              2:44:57
            </td>
            

          </tr>
          

          

          <tr
            class="TableRow"
            id="dk-table-12"
            pk="8"
            rownum="3"
          >
            

            <td
              axis="project"
              class="TableCell"
              id="dk-tr-17"
            >
              hGenerelt NT
            </td>
            

            

            <td
              axis="work"
              class="TableCell"
              id="dk-tr-18"
            >
              1:03:57
            </td>
            

          </tr>
          

          

          <tr
            class="TableRow"
            id="dk-table-13"
            pk="9"
            rownum="4"
          >
            

            <td
              axis="project"
              class="TableCell"
              id="dk-tr-19"
            >
              iTiktok
            </td>
            

            

            <td
              axis="work"
              class="TableCell"
              id="dk-tr-20"
            >
              2:44:57
            </td>
            

          </tr>
          

          

          <tr
            class="TableRow"
            id="dk-table-14"
            pk="10"
            rownum="5"
          >
            

            <td
              axis="project"
              class="TableCell"
              id="dk-tr-21"
            >
              jGenerelt NT
            </td>
            

            

            <td
              axis="work"
              class="TableCell"
              id="dk-tr-22"
            >
              1:03:57
            </td>
            

          </tr>
          

          

          <tr
            class="TableRow"
            id="dk-table-15"
            pk="11"
            rownum="6"
          >
            

            <td
              axis="project"
              class="TableCell"
              id="dk-tr-23"
            >
              kTiktok
            </td>
            

            

            <td
              axis="work"
              class="TableCell"
              id="dk-tr-24"
            >
              2:44:57
            </td>
            

          </tr>
          

          

          <tr
            class="TableRow"
            id="dk-table-16"
            pk="12"
            rownum="7"
          >
            

            <td
              axis="project"
              class="TableCell"
              id="dk-tr-25"
            >
              last
            </td>
            

            

            <td
              axis="work"
              class="TableCell"
              id="dk-tr-26"
            >
              1:06:43
            </td>
            

          </tr>
          

        </tbody>
      </table>
    </div>
    
                
    <div
      class="pgr"
    >
      <ul
        class="PagerWidget pagination"
        id="pager-widget-1"
        style="margin-top: 0px; margin-bottom: 0px;"
      >
        <li
          class="prev-nav"
        >
          <a
            href="#"
          >
            «
          </a>
        </li>
        <li
          class="page"
          page="1"
        >
          <a
            href="#"
          >
            1
          </a>
        </li>
        <li
          class="page active"
          page="2"
        >
          <a
            href="#"
          >
            2
          </a>
        </li>
        <li
          class="next-nav"
        >
          <a
            href="#"
          >
            »
          </a>
        </li>
      </ul>
    </div>
    
            
  </div>
  
        
</div>
`;
